src/sage/graphs/homomorphisms/count_homomorphisms.py

```diff
@@ -0,0 +1,327 @@
+
+        self.graph = graph
+        self.target_graph = target_graph
+        self.density_threshold = density_threshold
+        self.graph_clr = graph_clr
+        self.target_clr = target_clr
+        self.colourful = colourful
+
+        # Bookkeeping for colourful mappings
+        self.actual_target_graph = target_graph
+        self.actual_target_size = self.actual_target_graph.order()
+
+        self.graph._scream_if_not_simple()
+        self.target_graph._scream_if_not_simple()
+
+        self.tree_decomp = graph.treewidth(certificate=True)
+        self.nice_tree_decomp = make_nice_tree_decomposition(graph, self.tree_decomp)
+        self.root = min(self.nice_tree_decomp)
+        # self.root = sorted(self.nice_tree_decomp)[0]
+
+        # Make it into directed graph for better access
+        # to children and parent, if needed
+        #
+        # Each node in a labelled nice tree decomposition
+        # has the following form:
+        #
+        # (node_index, bag_vertices) node_type
+        #
+        # Example: (5, {0, 4}) intro
+        self.dir_labelled_TD = label_nice_tree_decomposition(self.nice_tree_decomp, self.root, directed=True)
+
+        # `node_changes_dict` is responsible for recording introduced and
+        # forgotten vertices in a nice tree decomposition
+        self.node_changes_dict = node_changes(self.dir_labelled_TD)
+
+        # `DP_table` is a list of lists of natural numbers (0 included)
+        # Each element (list) corresponds to the (induced) homomorphisms of
+        # a tree node.
+        self.DP_table = [] * self.dir_labelled_TD.order()
+
+    def count_homomorphisms(self):
+        r"""
+        Return the number of homomorphisms from the graph `G` to the graph `H`.
+
+        A homomorphism from a graph `G` to a graph `H` is a function
+        `\varphi : V(G) \mapsto V(H)`, such that for any edge `uv \in E(G)` the
+        pair `\varphi(u) \varphi(v)` is an edge of `H`.
+
+        For more information, see the :wikipedia:`Graph_homomorphism`.
+
+        ALGORITHM:
+
+        This is an implementation based on the proof of Prop. 1.6 in [CDM2017]_.
+
+        OUTPUT:
+
+        - an integer, the number of homomorphisms from `graph` to `target_graph`
+
+        EXAMPLES::
+
+            sage: from sage.graphs.homomorphisms.count_homomorphisms import GraphHomomorphismCounter
+            sage: square = graphs.CycleGraph(4)
+            sage: bip = graphs.CompleteBipartiteGraph(2, 4)
+            sage: counter = GraphHomomorphismCounter(square, bip)
+            sage: counter.count_homomorphisms()
+            128
+        """
+        # Whether it's BFS or DFS, every node below join node(s) would be
+        # computed already, so we can go bottom-up safely.
+        for node in reversed(list(self.dir_labelled_TD.breadth_first_search(min(self.dir_labelled_TD)))):
```

Comment:
Are you sure that `min(self.dir_labelled_TD)` is the root ? isn't it safer to search for a vertex with in-degree 0 ?

Reply:
> Are you sure that min(self.dir_labelled_TD) is the root ?

By construction, yes.